Store onboarding: analytics

jaclync opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #9248 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR implemented the analytics plan pe5sF9-15a-p2#comment-1933:

event name | timing | event properties
-- | -- | --
store_onboarding_shown | When the store onboarding card is shown on the dashboard. | We already track is_wpcom_store to know whether the store is hosted on wpcom.
store_onboarding_task_tapped | When the user taps on a task in the store onboarding card. | – task: launch_site \| products \| add_domain \| payments (general payments or WCPay) \| store_details(when we implement it)
store_onboarding_task_completed | When the user completes a task in the store onboarding card in the app (without confirming the onboarding/tasks API response since it might not be called if the user leaves the app). | – task: launch_site \| products \| add_domain (payment and store details tasks cannot be easily tracked since they are in a webview)
store_onboarding_completed | When the user completes all tasks in the store onboarding card (from the onboarding/tasks API response) and the card is being hidden.

<br class="Apple-interchange-newline">

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

### Site with all pending onboarding tasks

Prerequisite: a WPCOM site that already has a custom domain, and has 3 pending onboarding tasks - add product, launch site, payments

- Log in to the site as in the prerequisite --> when reaching the My store tab, the onboarding card should be shown and an event should be tracked `store_onboarding_shown` with a property `is_wpcom_store` of the expected value
- Tap on `Add your first product` --> an event should be tracked `🔵 Tracked store_onboarding_task_tapped, properties: [AnyHashable("task"): "products", AnyHashable("is_wpcom_store"): true, ...]`
- Proceed to create a product --> on completion, an event should be tracked `🔵 Tracked store_onboarding_task_completed, properties: [AnyHashable("is_wpcom_store"): true, AnyHashable("task"): "products", ...]`
- Go back to the My store tab
- Tap on `Launch your store` --> an event should be tracked `🔵 Tracked store_onboarding_task_tapped, properties: [AnyHashable("task"): "launch_site", AnyHashable("is_wpcom_store"): true, ...]`
- Tap `Publish My Store` --> on success, an event should be tracked `🔵 Tracked store_onboarding_task_completed, properties: [AnyHashable("is_wpcom_store"): true, AnyHashable("task"): "launch_site", ...]`
- Tap `Back to My Store`
- Tap on `Get paid` --> an event should be tracked `🔵 Tracked store_onboarding_task_tapped, properties: [AnyHashable("task"): "payments", AnyHashable("is_wpcom_store"): true, ...]`
- Tap on `Continue Setup` and add a payment method or set up WCPay in order to complete the task
- Go back to the My store tab --> the `Get paid` task should be complete, and an event should be tracked  `store_onboarding_completed`

### Site with pending custom domains task

Prerequisite: a WPCOM site that has at least a pending onboarding task on custom domains

- Log in to the site as in the prerequisite --> when reaching the My store tab, the onboarding card should be shown and an event should be tracked `store_onboarding_shown` with a property `is_wpcom_store` of the expected value
- Tap on `Customize your domain` --> an event should be tracked `🔵 Tracked store_onboarding_task_tapped, properties: [AnyHashable("is_wpcom_store"): true, AnyHashable("task"): "add_domain", ...]`
- Proceed to purchase a domain with store credit --> with the current behavior in web, the task isn't marked as complete until the domain is active about 30 minutes later

### Site without pending onboarding tasks

Prerequisite: a site that doesn't has any pending onboarding tasks

- Log in to the site as in the prerequisite --> even if the onboarding card is shown briefly in loading state, `store_onboarding_shown` event should **not** be tracked



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

jaclync committed:
Store onboarding analytics (plan to be confirmed).

jaclync committed:
Merge branch 'trunk' into feat/9248-onboarding-analytics

* trunk: (124 commits)
  Removes free trial mentiosn from release notes
  Update release notes
  Remove feature flag conditional
  Launch support from from upgrades view
  Improve copy
  Show Components row in product details for composite products
  Add composite products feature flag
  9202 Tidy up shouldShow logic on SetUpTTPCompleteVM
  9202 Use injected stores, not ServiceLocator
  Remove `XLPagerTabStrip` dependency from the Podfile.
  Delete unused files for `XLPagerTabStrip`.
  Add a comment on the multiline onboarding task subtitle size modifier.
  Dashboard: add 16px bottom margin to the store name label.
  Move the top vertical spacer to outside of the HStack so that the image and chevron icons can align vertically.
  Fix ambiguous number of lines for the onboarding task subtitle.
  Pass `onTaskCompleted` closure to fix tests.
  Store onboarding: remove spacing between tasks.
  Store onboarding: remove 16px bottom padding when `View all` CTA is not shown.
  Also shimmer the onboarding progress subtitle.
  Move `onDomainPurchased` to initializer.
  ...

# Conflicts:
#	WooCommerce/Classes/ViewRelated/Dashboard/Onboarding/StoreOnboardingCoordinator.swift
#	WooCommerce/Classes/ViewRelated/Dashboard/Onboarding/StoreOnboardingLaunchStoreCoordinator.swift
#	WooCommerce/Classes/ViewRelated/Dashboard/Settings/Domains/DomainSettingsCoordinator.swift

jaclync committed:
Track onboarding completed and shown events accurately.

jaclync committed:
Track onboarding task completed event in the existing closure.

jaclync committed:
Use the same analytics value for `payments` and `woocommercePayments` onboarding tasks.

itsmeichigo reviewed (approved):
All events tracked as expected ✅ 

👀 I saw the event `product_list_add_product_button_tapped` tracked when I tapped the create product task. Should this event be tracked only upon tapping the create product button instead of when the product creation form is presented?

Another observation that could be irrelevant to this PR: after publishing a site, I tapped the Back to my store button but the modal would not be dismissed. I had to drag to dismiss the modal, but this works in landscape mode.

<img src="https://user-images.githubusercontent.com/5533851/227106935-26867bed-2a84-4395-9edd-fbebf18a2f9a.png" width=320 />

jaclync committed:
Fix unit tests from `StoreOnboardingCoordinator.onTaskCompleted` interface changes.

jaclync committed:
Add source to `AddProductCoordinator` so that the product list Tracks event is not tracked from store onboarding.

jaclync commented:
> 👀 I saw the event `product_list_add_product_button_tapped` tracked when I tapped the create product task. Should this event be tracked only upon tapping the create product button instead of when the product creation form is presented?

great catch, this event shouldn't be tracked from store onboarding. I fixed this by creating a source enum for the coordinator and keeping the pre-existing tracking behavior in https://github.com/woocommerce/woocommerce-ios/pull/9268/commits/4bcaeb2b40e7dbaa41af5e85b56882ded0897eab

> Another observation that could be irrelevant to this PR: after publishing a site, I tapped the Back to my store button but the modal would not be dismissed. I had to drag to dismiss the modal, but this works in landscape mode.

thanks for reporting this! I added this as a subtask to https://github.com/woocommerce/woocommerce-ios/issues/9122.
